### ISSUE TITLE
Decoupling record type from table name

### ISSUE THREAD

**Issue by sepw** [groue/GRDB.swift#833]
I would like to use the query interface to build queries for my record type. However there isn't a 1:1 relationship between record type and underlying source/table:

- A record may come from one of 3 tables: `A`, `B` or `B`.
- The schemas of tables `A`, `B` and `C` are the same, i.e. they have the same number of columns and the columns have the same types.
- Table `A` is in the main database. Tables `B` and `C` may be in the temporary database or attached databases, either on disk or in memory.
- Records may be inserted into table `A`, `B` or `C` but fetched from view `V`, which is a UNION of `A`, `B` and `C`.

GRDB seems to assume that a record type comes from a single table, i.e. the table returned from `TableRecord. databaseTableName`. I need a way to decouple the table name from the type.

Is there a way to specify the table/view name when using the query interface? I was hoping for something like:

``` swift
_queue.read { db in
  return MyRecord.from("temp.B").fetchAll(db)
}
```

**Comment by sepw** [comment]
After some investigation into `TableRecord` and `QueryInterfaceRequest` it seems it might be straightforward to implement a `from(_ table: String)` function or similar as an extension on `TableRecord`, e.g.

``` swift
public static func from(_ table: String) -> QueryInterfaceRequest<Self> {
    let relation = SQLRelation(
        source: .table(tableName: table, alias: nil),
        selectionPromise: DatabasePromise(value: databaseSelection))
    return QueryInterfaceRequest(relation: relation)
}
```

Unfortunately `SQLRelation` has `internal` access control...

**Comment by groue** [comment]
Hello @sepw,

> GRDB seems to assume that a record type comes from a single table, i.e. the table returned from `TableRecord. databaseTableName`.

Exact.

> I need a way to decouple the table name from the type.

I have been waiting for this feature request :-)

> After some investigation [...] it seems it might be straightforward to implement.

Indeed. Enough years have passed now so I do feel rather confident that this should not create much problem. I'm all with you.

> `MyRecord.from("temp.B").fetchAll(db)`

This reads well, but `MyRecord.from("name").fetchAll(db)` does not. What is "name"?

`MyRecord.from(table: "name").fetchAll(db)` reads much better, but is not quite right with views.

Also, do we want to create such requests from `TableRecord`, which has its `databaseTableName` property? In your particular case, it looks like your record type does not have *one* natural source table, but several. I wouldn't like to force users to "bless" one of the tables. I would even like to force users to pick one of those table, explicitly, for each request.

What do you think?

**Comment by sepw** [comment]
Hi Groue,

Many thanks responding so quickly!

> This reads well, but MyRecord.from("name").fetchAll(db) does not. What is "name"?
> 
> MyRecord.from(table: "name").fetchAll(db) reads much better, but is not quite right with views.

I personally don't find the absence of the first parameter name to be a problem. In fact, `from("source")` is so close to the underlying SQL that it's likely to be clear what `"source"` is. It also side-steps the issue of "what if the source is not a table"? ;-)

> Also, do we want to create such requests from `TableRecord`, which has its `databaseTableName` property? In your particular case, it looks like your record type does not have one natural source table, but several. I wouldn't like to force users to "bless" one of the tables. I would even like to force users to pick one of those table, explicitly, for each request.

In my case I want to implement a "cascade" of data sets from a set of "domains". Think of it like CSS, where styles cascade from browser-defined stylesheet(s), to author-defined stylesheet(s) to user-defined stylesheet(s). My user-defined records are stored in the `main` database on disk. This set is then augmented with additional sets of system-defined records, likely in attached in-memory databases. A view would then be used to encapsulate the aggregation of records from these various sources.

So, while the records come from a number of different sources, it would nevertheless be appropriate to "bless" the table in `main` as the primary table associated with the record type. The use of `from()` would simply override the table defined by the record type's `databaseTableName`.

(On a side-note: `MyRecord.from("table").fetchAll(db)` reads well for fetches, but wouldn't work for inserts and updates, which would probably benefit from alternatives such as `into()` and `in()`...)

Many thanks for your input,
Simon

**Comment by sepw** [comment]
It occurred to me that adding a prefix to the table name using the proposed `from()` function will probably be insufficient. It appears that GRDB transforms table names into quoted identifiers when formatting SQL statements. This code will need to be aware of schema prefixes in order that a table table can be quoted correctly, i.e.

```
schema.table
```

needs to be quoted as

```
"schema"."table"
```

and not

```
"schema.table"
```

I've taken a quick look at the GRDB code. As far as I can tell the formatting of table names seems to be handled by `SQLQualifiedSource` (specifically `SQLQualifiedSource.sql(_:)`), which in turn delegates to the `quotedDatabaseIdentifier` extension on `String` in `Utils.swift`. This would probably need to be changed from

``` swift
@inlinable public var quotedDatabaseIdentifier: String {
    // See https://www.sqlite.org/lang_keywords.html
    return "\"\(self)\""
}
```

to something like

``` swift
@inlinable public var quotedDatabaseIdentifier: String {
    // See https://www.sqlite.org/lang_keywords.html
    return components(separatedBy: ".").map { "\"\($0)\"" }.joined(separator: ".")
}
```

**Comment by groue** [comment]
Hello again @sepw,

I've been thinking about this issue a little more, and I'm less comfortable than before with the idea of overriding the default table name willy-nilly.

First, the fact that `TableRecord` is tied to a single table is actually used in many places of the GRDB implementation. For example, associations, primary-key-based request modifiers, persistence methods (and particularly updates), table introspection, all those features need a table name. None of those features were prepared for a last-minute table change. There is a fair amount of necessary work and testing. Also think that nothing prevents the user from substituting one table with another that has the same layout, but different keys, indexes, or even primary key.

I mean that adding one method is rarely enough. I want to make sure it blends correctly in the rest of the API surface.

Second, thank you for describing you use case of "cascading domains implemented with attached databases".

It had me think a little bit, and in particular I imagined a natural extension to it, which is the desire of the user to remember which table/attached table a particular record value comes from/should be persisted to. I didn't see how your suggested apis can handle such extension. To me this is the sign that we haven't yet well understood what we're after. Well, maybe *you* do. But as an API maintainer, I don't want to address one use case from one person who works on one particular application. Such incomplete work belongs to application code, or a fork.

I have two questions for you, which should help things moving forward:

- How do you currently deal with your "cascading domains", in the current state of the GRDB api?
- How many tables does your cascading record type have to feed from? A known and limited set of tables and views, or an unknown and dynamic set of tables and views which evolves at runtime?

> (On a side-note: MyRecord.from("table").fetchAll(db) reads well for fetches and deletes, but wouldn't work for inserts and updates, which would probably benefit from alternatives such as into() and in()...)

Duly noted, thanks :-)

> It appears that GRDB transforms table names into quoted identifiers when formatting SQL statements. This code will need to be aware of schema prefixes in order that a table table can be quoted correctly, i.e. `schema.table` needs to be quoted as `"schema"."table"` and not `"schema.table"`.

That's a very useful contribution, thank you 👍 I did not spend much time on attached databases and alternative database names. I'll surely have a look at this shortly!

**Comment by sepw** [comment]
> I have two questions for you, which should help things moving forward:
> 
> - How do you currently deal with your "cascading domains", in the current state of the GRDB api?

I am currently designing the storage sub-system, but these are the parts of GRDB I would very much like to be able to use:

- `DatabaseMigrator` along with the Swift-based DDL commands on `Database` e.g. `create(table:...)` etc.
- `Codable` record types -- I would prefer to avoid using `Row` instances directly.
- Query Interface -- I am very familiar with SQL and SQLite, but I would prefer not to have SQL literals littered about in my code ;-)

> - How many tables does your cascading record type feed from? A known and limited set of tables and views, or an unknown and dynamic set of tables and views which evolves at runtime?

A fixed set of tables in:

1. The primary, persistent database `main`
2. A second ephemeral set, either in `temp` or an attached in-memory database.
3. Possibly a third ephemeral set in an in-memory database, although this is currently unclear.

> I've been thinking about this issue a little more, and I'm less comfortable than before with the idea of overriding the default table name willy-nilly.

I totally understand your reservations. It wasn't my intention to get something like this squeezed into GRDB 5 at the 11th hour! Instead I was more asking if it would be technically feasible for me to extend GRDB using extensions to `QueryInterfaceRequest<>` in *my* codebase (hence my observation about `SQLRelation` having internal access)...

I have also given this more thought. I spent the morning exploring GRDB's current compatibility with qualified names, e.g. `schema.table` rather than just `table`.

The first thing that I tried was `Database.create(table:..)` and this didn't work due to limitations with the identifier quoting I mentioned previously. I also tried to use record types that return a schema-prefixed name from `databaseTableName` e.g.

``` swift
struct MyRecord: TableRecord {
  static let databaseTableName: "attached.MyTable"
}
```

Unfortunately this also didn't work.

So I forked GRDB to see how straightforward it would be to add support for qualified names. I've invested a couple of hours this morning and have made some progress. I have record-based `insert()`, `fetchXXX()`, `update()` and `deleteXXX()` working. I also have `Database.create(table:...)` functioning. I am currently looking at `Database.tableExists()`.

I'd like to emphasise that my focus here is *not* the query interface extensions I originally described: I'm just trying to get GRDB to work with qualified table names e.g. `schema.table` to allow me to create/drop tables, determine whether tables exist etc...

Would you be interested in me continuing this work? GRDB is a great project and I think it would benefit from supporting qualified names. I'd like to contribute if I can, although the resources I have at my disposal are pretty limited...

Many thanks

**Comment by groue** [comment]
@sepw, I'll gladfully accept a pull request which adds this (and only this), namely the ability to pass qualified names where table names are expected. You already started listing such places:

- `TableRecord.databaseTableName`
- [Schema methods](https://github.com/groue/GRDB.swift#database-schema)
- [Introspection methods](https://github.com/groue/GRDB.swift#database-schema-introspection)
- [SQL Interpolation](https://github.com/groue/GRDB.swift/blob/master/Documentation/SQLInterpolation.md#sql-interpolation-reference)

That would be really great!

Back on the topic of records that read and write into multiple tables.

Your efforts to have `insert()`, `fetchXXX()` etc work in your app will certainly be successful. I just worry they eventually hit the `TableRecord` limitations I talked about above, since this protocol is tied with the assumption that it targets a single table. Forking GRDB without any hope of having your changes merged upstream is... a hard bet for the future and maintenance of your app.

An option we haven't tried yet, since your apps uses a fixed set of tables, is to define one record type per table, and see what it takes to make them ergonomic enough for your app to use them.

**Comment by sepw** [comment]
> Your efforts to have insert(), fetchXXX() etc work in your app will certainly be successful. I just worry they eventually hit the TableRecord limitations I talked about above, since this protocol is tied with the assumption that it targets a single table. Forking GRDB without any hope of having your changes merged upstream is... a hard bet for the future and maintenance of your app.

I have no intention of using a forked version of GRDB! I only forked it to explore the feasibility of adding support for qualified names.

> An option we haven't tried yet, since your apps uses a fixed set of tables, is to define one record type per table, and see what it takes to make them ergonomic enough for your app to use them.

I have considered internally using a record type that is generic over a schema type that provides the name of the schema the record should use:

``` swift
protocol SchemaTraits {
  static var name: String { get }
}

struct MainDatabase: SchemaTraits {
  public static let name { "main" }
}

struct AttachedDatabase: SchemaTraits {
  public static let name { "attached" }
}

struct MyRecord<Schema: SchemaTraits>: TableRecord, FetchableRecord, PersistableRecord {
  public static var databaseTableName: String{ Schema.name }
  ...
}
```

I could then use MyRecord<MainDatabase> or MyRecord<AttachedDatabase> with the GRDB record layer. It's not perfect as I would need to keep these types internal and use a separate public type for my API, but it would work.

However, it seems like the payoff off for using the record/query interfaces would be too small considering the impedance mismatch between what I'm trying to do and what GRDB supports. I would probably be better off dropping down to the SQL layer or pursuing a different data architecture.

> I'll gladfully accept a pull request which adds this (and only this), namely the ability to pass qualified names where table names are expected. You already started listing such places.

I have invested about 4 hours into adapting GRDB to support qualified names. I have the inserts, fetches, updates and deletes working for records. Table create, alter and drop on `Database` also work. I have also adapted all methods in `Database+Schema.swift`.

Most changes were minor, with the only stumbling blocks being `viewExists()` and `triggerExists()` as schema objects from attached databases are not present in `sqlite_master`. I had to resort to using `PRAGMA table_info` for `tableExists()`.

I have created 30-40 unit tests to test these features both with `temp` and attached databases.

But at this point I don't think it makes sense for me to continue much further:

- I don't think just adding support for qualified table names make sense for GRDB. If support for qualified names is added then it needs to be added for all object types including views, triggers...
- GRDB has a very large surface area. This makes it very difficult for me to estimate how much needs to be done to complete the work I have started. But my gut feeling is that it isn't feasible for me to make the necessary changes in the time I have available.
- If I can't use the record and query interface layers of GRDB (see above) and I am going to have to drop down to SQL then I don't actually need support for qualified names in the rest of the API.

What should I do with the changes I have made? I can submit a pull request to you if you would like but I can't really gauge how complete the work is. I don't think it makes sense to merge these changes into your development stream but it might be useful to you in the future if you parked the code in a branch for future reference?

**Comment by groue** [comment]
I 100% understand your reasoning.

I'm not sure it is worth asking more help from you: you already described really well the landscape, and that's quite valuable. The difficulty with qualified tables is that the GRDB API has grown without any consideration for them and attached database in general. And we have learned that just adding some quick fixes... kind of misses the point anyway.

I don't quite know how I can provide further help, and this somewhat troubles me. It's uncommon that the library really can't address an application need 😅 

So... If you think I can do something, please ask.

**Comment by sepw** [comment]
Hi @groue

> I don't quite know how I can provide further help, and this somewhat troubles me. It's uncommon that the library really can't address an application need

No worries! GRDB is a great library and I'm currently looking at alternative data architectures that will let me take advantage of its features to the full.

**Comment by groue** [comment]
All right, let's close this issue, then. It's been a great exploration. And you're very welcome!

**Comment by wzio** [comment]
I encountered a similar problem
Imagine an instant messaging app
with a list of sessions and a list of messages
Suppose there are two ways:
1 One way is to design a session table and a message table, where all group messages are stored in one message table
2 The other way is to design a session table and then a message table for each session. so the message table name may be related to the group id, but the `databaseTableName` in the library cannot be dynamic

Also if someone wants to split the table by date, they will encounter a similar problem

**Comment by groue** [comment]
Hello @wzio,

Encoding data in the database schema is a bad practice, because it makes it impossible to leverage some relational constraints. For example, you can't make sure that for each message table, one matching row exists in the session table, or that all messages for a given session are deleted when the session is deleted - you can easily end up with dangling messages (see [foreign key constraints](https://sqlite.org/foreignkeys.html)). Splitting by date comes with even more difficulties, since you can no longer perform queries across dates without an immense pain.

My advice for you is to focus on the "normal setup" of a database, which is a schema with a known and fixed set of tables. Strengthen your knowledge of relational databases so that you use the features of SQLite in order to guarantee the integrity of your data (words like primary, secondary, or foreign keys, indexes, check constraints, etc should enter your vocabulary). You will avoid a whole class of bugs in your app. The schema of your SQLite database is more important than the Swift code that accesses it.

I do not mean that schemas with a variable set of tables are forbidden. But before doing so, it is necessary to describe why the plain relational model has reached a limit. I'm not sure your messaging app is such a case. Sharding can be useful, but it has a high complexity and maintenance cost.

**Comment by wzio** [comment]
Hi, @groue 
Thanks for your advice.
A conservative app like WhatsApp, Messenger may generate 300 messages per day, 100,000 messages per year, if used more frequently, I think a huge message table is not recommended. so these apps have to divided into tables use a same schema. In this case,  a feature like `Message.from(table: "group_id_xxx").fetchAll(db)` will be good

**Comment by groue** [comment]
All right, @wzio! I can't wait for an engineer from WhatsApp to contact me, and provide benchmarks that support the need for sharding. SQLite supports _huge_ databases without performance loss, as Google searches will tell you.

**Comment by groue** [comment]
QueryInterfaceRequest builder with dynamic table/view name is addressed by the pull request #993.
